File: backend/app/schemas/memory.py
"""API request/response models for the v2 memory subsystem."""

from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field

from app.schemas.contracts import ClaimData


class WriteFactRequest(BaseModel):
    claim: ClaimData
    session_id: str | None = None
    lane: str | None = None


class WriteFactResponse(BaseModel):
    status: str
    lane: str
    claim_id: str | None = None
    reason: str | None = None
    rebound_active: bool = False
    conflict: bool = False


class RecallResponse(BaseModel):
    text: str
    identity: list[str]
    relevant: list[str]
    related: list[str]
    rules: list[str]


class SnippetRequest(BaseModel):
    role: str
    content: str
    session_id: str | None = None


class SnippetResponse(BaseModel):
    snippet_id: str
    role: str
    content: str
    session_id: str | None
    created_at: datetime
    expires_at: datetime | None


class EntityRequest(BaseModel):
    name: str
    entity_type: str | None = None
    attributes: dict[str, Any] | None = None


class EntityResponse(BaseModel):
    entity_id: str
    name: str
    entity_type: str | None
    attributes: dict[str, Any] | None


class RelationRequest(BaseModel):
    subject_id: str
    predicate: str
    object_id: str
    confidence: float = 0.5


class RelationResponse(BaseModel):
    relation_id: str
    subject_id: str
    predicate: str
    object_id: str
    confidence: float


class ProposeRuleRequest(BaseModel):
    behavior_text: str
    domain: str = "global"
    trigger: dict[str, Any] = Field(default_factory=dict)
    effect: dict[str, Any] = Field(default_factory=dict)
    artifact_cost: int = 1
    source_type: str = "user_explicit"


class RuleConflictResponse(BaseModel):
    conflict_id: str
    rule_id: str
    other_rule_id: str | None
    conflict_type: str
    severity: str
    detail: str | None
    resolved: bool


class RuleResponse(BaseModel):
    rule_id: str
    domain: str
    behavior_text: str
    trigger: dict[str, Any]
    effect: dict[str, Any]
    artifact_cost: int
    status: str
    source_type: str
    previous_rule_id: str | None
    created_at: datetime
    activated_at: datetime | None
    expires_at: datetime | None


class ProposeRuleResponse(BaseModel):
    rule: RuleResponse
    conflicts: list[RuleConflictResponse]
    has_hard_conflict: bool


class ApproveRuleRequest(BaseModel):
    override_soft: bool = False


class RejectRuleRequest(BaseModel):
    reason: str | None = None


class ProvenanceEventResponse(BaseModel):
    event_type: str
    result: str
    actor: str
    detail: str
    timestamp: str | None


class MemoryStatsResponse(BaseModel):
    claims_by_lane: dict[str, int]
    open_conflicts: int
    entities: int
    relations: int
    snippets: int
    behavioral_rules_active: int
    behavioral_rules_pending: int

File: frontend/src/api/memory.ts
import { request } from "@/api/client";

export type RuleConflict = {
  conflict_id: string;
  rule_id: string;
  other_rule_id: string | null;
  conflict_type: "direct" | "interaction" | "budget" | "cap";
  severity: "hard" | "soft";
  detail: string | null;
  resolved: boolean;
};

export type BehavioralRule = {
  rule_id: string;
  domain: string;
  behavior_text: string;
  trigger: Record<string, unknown>;
  effect: Record<string, unknown>;
  artifact_cost: number;
  status: "pending" | "active" | "rejected" | "retired";
  source_type: string;
  previous_rule_id: string | null;
  created_at: string;
  activated_at: string | null;
  expires_at: string | null;
};

export type MemoryStats = {
  claims_by_lane: Record<string, number>;
  open_conflicts: number;
  entities: number;
  relations: number;
  snippets: number;
  behavioral_rules_active: number;
  behavioral_rules_pending: number;
};

export function listBehavioralRules(status?: string): Promise<BehavioralRule[]> {
  const query = status ? `?status=${encodeURIComponent(status)}` : "";
  return request<BehavioralRule[]>(`/memory/rules${query}`);
}

export function listRuleConflicts(ruleId?: string): Promise<RuleConflict[]> {
  const query = ruleId ? `?rule_id=${encodeURIComponent(ruleId)}` : "";
  return request<RuleConflict[]>(`/memory/rules/conflicts${query}`);
}

export function approveBehavioralRule(id: string, overrideSoft = false): Promise<BehavioralRule> {
  return request<BehavioralRule>(`/memory/rules/${id}/approve`, {
    method: "POST",
    body: { override_soft: overrideSoft },
  });
}

export function rejectBehavioralRule(id: string, reason?: string): Promise<BehavioralRule> {
  return request<BehavioralRule>(`/memory/rules/${id}/reject`, {
    method: "POST",
    body: reason ? { reason } : {},
  });
}

export function retireBehavioralRule(id: string, reason?: string): Promise<BehavioralRule> {
  return request<BehavioralRule>(`/memory/rules/${id}/retire`, {
    method: "POST",
    body: reason ? { reason } : {},
  });
}

export function getMemoryStats(): Promise<MemoryStats> {
  return request<MemoryStats>("/memory/stats");
}
